clang/lib/Sema/SemaOpenACC.cpp

```diff
@@ -2627,16 +2664,40 @@ static Expr *GenerateReductionInitRecipeExpr(ASTContext &Context,
   } else {
     assert(Ty->isScalarType());
 
-    // TODO:  OpenACC: This currently only works for '1', but we need to figure
-    // out a way to do least/largest/all-1s.
-    if (Ty->isFloatingType()) {
-      Exprs.push_back(FloatingLiteral::Create(
-          Context, llvm::APFloat::getOne(Context.getFloatTypeSemantics(Ty)),
-          /*isExact=*/true, Ty, ExprRange.getBegin()));
+    if (const auto *Cplx = Ty->getAs<ComplexType>()) {
+      // we can get here in error cases, so make sure we generate something that
+      // will work if we find ourselves wanting to enable this.
+
+      QualType EltTy = Cplx->getElementType();
+      if (EltTy->isFloatingType()) {
+        Exprs.push_back(FloatingLiteral::Create(
+            Context, getInitFloatValue(Context, IK, EltTy),
+            /*isExact=*/true, EltTy, ExprRange.getBegin()));
+        Exprs.push_back(FloatingLiteral::Create(
```

Comment:
This probably isn't what you want. For example, for `InitKind::One` this will generate `1+1i` but the identity value for complex multiplication is `1 + 0i`. Maybe best to leave it as an error-producing case.

Reply:
Could you guide me as to what the right answer is for them?  I could definitely split these up into separate functions, but I'd love to know what the '0', '1', All-Ones, Least and Largest should be here?

I think I'd one day soon want to enable complex here.

Reply:
Ah, hmm.. ok then, I think leaving complex as ill-formed is the best idea. 

I DID just realize my test before was wrong, and I HAVEN'T been disallowing complex, so I'm going to go through and do that as a RAC patch.

Reply:
Oops. Magnitude is an absolute value, so I guess you'd need {0, 0} for the LEAST case. Anyway, it's probably just a bad idea to allow it at all.

Reply:
I could probably make complex legal for some operators (+/*?), but it doesn't seem worth the effort until someone asks for it.  In the meantime, I've made sure it is ill-formed always, here: 7a0dfb1fec20bcb8565b39507e282b12e436ba1f